packages/block-editor/src/components/inserter/block-list.js

```diff
@@ -135,6 +135,10 @@ function InserterBlockList( {
 		return filter( filteredItems, { category: 'reusable' } );
 	}, [ filteredItems ] );
 
+	const uncategorizedItems = useMemo( () => {
```

Comment:
I think the pattern we've arrived at with various filtering on blocks has grown to a point where it could probably do for a refactoring to generalize the approach. I wasn't very keen to do it here, since it would inevitably touch on a few other problems with the implementation very much outside of scope (#22279, https://github.com/WordPress/gutenberg/pull/17609#issuecomment-626938994, https://github.com/WordPress/gutenberg/pull/17609#issuecomment-626944735). In the interim, I've merely adopted the established pattern.

Reply:
Should we include also the items that have a "category" but that category is not registered?

Reply:
> Should we include also the items that have a "category" but that category is not registered?

It works this way (indirectly) as a result of the additional changes in `registration.js`, where the `settings.category` is unset if it had been set to a category which is not registered.

Reply:
The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

Reply:
> The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

Maybe it's something where the `blockTypes` reducer should handle this behavior, including in response to category-affecting action types? I guess the downside is that if an earlier registered block type category had been considered invalid but is now _valid_, it wouldn't be restored.

Another option could be that this is handled somewhere in a selector like `getBlockType`.

In the end, I think it depends if we'd rather have certainty in knowing that data "at rest" is valid (or at least valid from the selector access point), or if we want to push this onto the specific UI to handle. The latter seems it could be cumbersome in how essentially it would need access to all categories just to determine if the category for a particular block type is valid.

Reply:
Since these selectors don't change that often (unlikely to cause performance issues), I personally think the latter approach is probably the best one. Memoization does have some costs though.

Reply:
> The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

I also wonder how much of a problem this is in practice, and whether it's something where we can be content in saying "the value is validated at the time of registration", irrespective of what happens after that point in time. In a sense, this is consistent with how block registration works in other ways. We currently forbid a block from being registered if its category doesn't exist. We never previously cared if that category _came into existence_ after that attempt at registration.

I'm fine any of these ways, to be honest. The selector seems like the best middle-ground, but as you note, there's some trade-offs to consider here, and a slight extra complexity in needing to account for this outside of registration-time considerations.

Reply:
I agree that it's not a big concern and I'm happy to move forward as is in this PR.

I'm thinking more as an ideal world here:

 - no usage of `wp.blocks.*` APIs.
 - rely 100% on selectors and actions.
 - registration timing not impactful.

There's a larger goal/dream here and just trying to see how we can fit it here :).

Reply:
Yeah, it sounds nice. I started to consider another alternative to the selector, which is to have a second "internal" state which maintains all initial block registrations, irrespective of their validity. It would be useful for categories, to solve what I described in https://github.com/WordPress/gutenberg/pull/22280#discussion_r423734129 as the problem of "an earlier registered block type category had been considered invalid but is now valid, it wouldn't be restored". I could see how something like this could be useful for other types of validations which can change over time. Looking at the [registration logic](https://github.com/WordPress/gutenberg/blob/89e21cf632a352c38844e12a34cec22bfefd12a6/packages/blocks/src/api/registration.js#L142-L242), I'm not sure anything else would fall under this.

I'd probably agree it's something we could consider addressing separately.